hamt/hamt.go

```diff
@@ -419,75 +434,109 @@ func (ds *Shard) walkTrie(ctx context.Context, cb func(*Shard) error) error {
+func (ds *Shard) setValue(ctx context.Context, hv *hashBits, key string, value *ipld.Link) (oldValue *ipld.Link, err error) {
 	idx, err := hv.Next(ds.tableSizeLg2)
 	if err != nil {
-		return err
+		return
 	}
 
 	if !ds.childer.has(idx) {
-		return ds.childer.insert(key, val, idx)
+		// Entry does not exist, create a new one.
+		return nil, ds.childer.insert(key, value, idx)
 	}
 
 	i := ds.childer.sliceIndex(idx)
-
 	child, err := ds.childer.get(ctx, i)
 	if err != nil {
-		return err
+		return
 	}
 
 	if child.isValueNode() {
+		// Leaf node. This is the base case of this recursive function.
+		// FIXME: Misleading: the base case also includes a recursive call
+		//  in the case both keys share the same slot in the new child shard
+		//  below.
 		if child.key == key {
-			// value modification
-			if val == nil {
-				return ds.childer.rm(idx)
+			// We are in the correct shard (tree level) so we modify this child
+			// and return.
+			oldValue = child.val
+
+			if value == nil { // Remove old entry.
+				return oldValue, ds.childer.rm(idx)
 			}
 
-			child.val = val
-			return nil
+			child.val = value // Overwrite entry.
+			return
 		}
 
-		if val == nil {
-			return os.ErrNotExist
+		if value == nil {
+			return nil, os.ErrNotExist
+			// FIXME: Move this case to the top of the clause as
+			//  if child.key != key
+			//  and remove the indentation on the big if.
 		}
 
-		// replace value with another shard, one level deeper
-		ns, err := NewShard(ds.dserv, ds.tableSize)
+		// We are in the same slot with another entry with a different key
+		// so we need to fork this leaf node into a shard with two childs:
+		// the old entry and the new one being inserted here.
+		// We don't overwrite anything here so we keep:
+		//   `oldValue = nil`
+
+		// The child of this shard will now be a new shard. The old child value
+		// will be a child of this new shard (along with the new value being
+		// inserted).
+		grandChild := child
+		child, err = NewShard(ds.dserv, ds.tableSize)
 		if err != nil {
-			return err
+			return nil, err
 		}
-		ns.builder = ds.builder
+		child.builder = ds.builder
 		chhv := &hashBits{
-			b:        hash([]byte(child.key)),
+			b:        hash([]byte(grandChild.key)),
 			consumed: hv.consumed,
 		}
 
-		err = ns.modifyValue(ctx, hv, key, val)
+		// We explicitly ignore the oldValue returned by the next two insertions
+		// (which will be nil) to highlight there is no overwrite here: they are
+		// done with different keys to a new (empty) shard. (At best this shard
+		// will create new ones until we find different slots for both.)
+		// FIXME: These two shouldn't be recursive calls where one function
+		//  adds the child and the other replaces it with a shard to go one
+		//  level deeper. This should just be a simple loop that traverses the
+		//  shared prefix in the key adding as many shards as needed and finally
+		//  inserts both leaf links together.
```

Comment:
Eh, is it really an issue? This code is naturally recursive and has a max depth of the number of bytes in the hash.

Reply:
Removed.